### PR TITLE
refactor: define api register of supernode

### PR DESCRIPTION
**Pull request by lowzj**
Signed-off-by: lowzj <zj3142063@gmail.com>

### Ⅰ. Describe what this PR did
 define and expose the api register of supernode to make it possible to add APIs outside.

### CODE REVIEW

File: supernode/server/api/api.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	}
+}
+
+// RegisterLegacy is deprecated, just for compatibility with the old version,
+// please do not use it to add new API.
+func RegisterLegacy(h *HandlerSpec) {
+	registerAPI(LegacyCategory, h)
+}
+
+func registerAPI(c *Category, h *HandlerSpec) {
```
Comment (wangforthinker) [inline supernode/server/api/api.go:53]: How about to allow register user-defined Category,  may be we want to register more than one groups of API.

File: supernode/server/api/utils.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/dragonflyoss/Dragonfly/apis/types"
+	"github.com/dragonflyoss/Dragonfly/pkg/errortypes"
+)
+
+// ValidateFunc validates the request parameters.
+type ValidateFunc func(registry strfmt.Registry) error
+
+// ParseJSONRequest parses the request parameter formed by JSON to a object.
+func ParseJSONRequest(body io.Reader, request interface{}, validator ValidateFunc) error {
+	if err := json.NewDecoder(body).Decode(request); err != nil {
```
Comment (wangforthinker) [inline supernode/server/api/utils.go:35]: Add validation of "body == nil " and "request == nil" ?

File: supernode/server/router.go [hunk 3]
```diff
@@ -22,11 +22,12 @@ import (
 	"net/http"
 	"net/http/pprof"
 
-	"github.com/dragonflyoss/Dragonfly/apis/types"
-	"github.com/dragonflyoss/Dragonfly/version"
 	"github.com/gorilla/mux"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	"github.com/dragonflyoss/Dragonfly/apis/types"
```
Comment (wangforthinker) [inline supernode/server/router.go:29]: The self-project import path may be better in second  part.

File: supernode/server/api/api.go (hunk 1 above)
Comment (lowzj) [reply supernode/server/api/api.go:53]: I think about it again, it's better to provide an extension category for users to register APIs. Users can groups their own customized APIs in this extension category, not in Dragonfly core APIs category.
WDYT?

File: supernode/server/api/utils.go (hunk 2 above)
Comment (lowzj) [reply supernode/server/api/utils.go:35]: It should return `400` response in this case.

File: supernode/server/router.go (hunk 3 above)
Comment (lowzj) [reply supernode/server/router.go]: done

File: supernode/server/api/api.go (hunk 1 above)
Comment (lowzj) [reply supernode/server/api/api.go:53]:
```json
GET /api/ext
```
And supernode provides the above API to list the current registered APIs.

File: supernode/server/api/handler.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+package api
+
+import (
+	"context"
+	"net/http"
+)
+
+// HandlerSpec describes a HTTP api
```
Comment (Starnop) [inline supernode/server/api/handler.go:24]: s/a/an

File: supernode/server/api/api.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+ * limitations under the License.
+ */
+
+package api
+
+var handlerCategories map[string][]*HandlerSpec
+
+const (
+	// VersionedPrefix is recommended, any new API should start with this prefix.
+	VersionedPrefix = "/api/v1"
```
Comment (Starnop) [inline supernode/server/api/api.go:23]: s/VersionedPrefix/V1VersionedPrefix
Maybe we will add v2,v3  API version in the future. 😄

**Comment by Starnop** [comment]
LGTM.

**Pull request merged**